### PR TITLE
use connBag to pass connections

### PR DESCRIPTION
**Pull request by fffw**

### CODE REVIEW

**Comment by fffw** [comment]
Fixes #45

File: src/github.com/getlantern/http-proxy/server.go [hunk 1]
```diff
@@ -164,12 +183,15 @@ func (s *Server) doServe(listener net.Listener, chListenOn *chan string) error {
...
+				cb.Put(c)
+			case http.StateClosed:
+				// When go server encounters abnormal request, it
+				// will transit to StateClosed directly without
+				// the connection being withdrawed.
+				// Purge it in such case.
+				cb.Purge(c.RemoteAddr().String())
 			}
 
 			if atomic.LoadUint64(&s.numConns) >= s.maxConns {
```
Comment (fffw) [inline src/github.com/getlantern/http-proxy/server.go]: BTW, I found the code executes during each state transition. If we only want to limit connections, we can run it when `StateNew`.

Comment (uaalto) [inline src/github.com/getlantern/http-proxy/server.go]: I didn't know the existence of StateNew. Yes, it's better.

File: src/github.com/getlantern/http-proxy/server.go [hunk 2]
```diff
@@ -164,12 +183,15 @@ func (s *Server) doServe(listener net.Listener, chListenOn *chan string) error {
...
-			if state == http.StateActive {
-				select {
-				case q <- c:
-				default:
-					fmt.Print("Oops! the connection queue is full!\n")
-				}
+			switch state {
+			case http.StateActive:
+				cb.Put(c)
+			case http.StateClosed:
```
Comment (uaalto) [inline src/github.com/getlantern/http-proxy/server.go]: Hijacked connections will never transition to this state, so those will never be purged.
We should also make sure the state changes to close with connection errors and potential panics in the handler goroutines. While this method is cleaner than the previous circular buffer, it has the potential to leak if we don't take precautions.

Comment (fffw) [inline src/github.com/getlantern/http-proxy/server.go]: I may chose a wrong name, but `connBag.Withdraw()` which invoked as the very first statement in handler will also remove it. `Purge` here only takes effect when the handler didn't get called.

File: src/github.com/getlantern/http-proxy/server.go (hunk 1 above)
Comment (fffw) [inline src/github.com/getlantern/http-proxy/server.go]: Ok. As we both agree on this, I updated code directly.

Comment (uaalto) [inline src/github.com/getlantern/http-proxy/server.go]: Yes, that's fine. We won't have conflicts merging :)

**Comment by fffw** [comment]
Oh sorry @uaalto I didn't notice your changes on master. Merged them now.

File: src/github.com/getlantern/http-proxy-extensions/mimic_test.go [hunk 3]
```diff
@@ -18,7 +18,7 @@ import (
 
 const target = "test-data/apache-2.4.7-ubuntu14.04.raw"
 const template = "test-data/apache-2.4.7-ubuntu14.04.tpl"
-const current = "test-data/chained-server.raw"
+const current = "test-data/http-proxy.raw"
```
Comment (uaalto) [inline src/github.com/getlantern/http-proxy-extensions/mimic_test.go]: Sorry, I changed that in the .gitignore file. I thought it was automatically generated in the build process, derived from the binary name!

**Comment by uaalto** [comment]
LGTM!

**Pull request merged**